fs/remote/blob_test.go

```diff
@@ -278,40 +280,213 @@ func TestFailReadAt(t *testing.T) {
 		t.Errorf("must be fail for broken multipart body but err=%v (allowMultiRange=%v)", err, allowMultiRange)
 		return
 	}
 }
 
 func checkBrokenHeader(t *testing.T, allowMultiRange bool) {
-	r := makeBlob(t, int64(len(sampleData1)), sampleChunkSize, defaultPrefetchChunkSize, brokenHeaderRoundTripper(t, []byte(sampleData1), allowMultiRange))
+	r := makeTestBlob(t, int64(len(sampleData1)), sampleChunkSize, defaultPrefetchChunkSize, brokenHeaderRoundTripper(t, []byte(sampleData1), allowMultiRange))
 	respData := make([]byte, len(sampleData1))
 	if _, err := r.ReadAt(respData[0:len(sampleData1)/2], 0); err == nil || err == io.EOF {
 		t.Errorf("must be fail for broken multipart header but err=%v (allowMultiRange=%v)", err, allowMultiRange)
 		return
 	}
 }
 
-func makeBlob(t *testing.T, size int64, chunkSize int64, prefetchChunkSize int64, fn RoundTripFunc) *blob {
-	return &blob{
-		fetcher: &fetcher{
+func TestParallelDownloadingBehavior(t *testing.T) {
+	type testData struct {
+		name           string
+		regions        [][]region
+		roundtripCount int64
+	}
+
+	tests := []testData{
+		{
+			name:           "no_data",
+			regions:        make([][]region, 0),
+			roundtripCount: 0,
+		},
+		{
+			name: "same_regions",
+			regions: [][]region{
+				{
+					{
+						b: 0,
+						e: 512000,
+					},
+				},
+				{
+					{
+						b: 0,
+						e: 512000,
+					},
+				},
+				{
+					{
+						b: 0,
+						e: 512000,
+					},
+				},
+			},
+			roundtripCount: 1,
+		},
+		{
+			name: "same_regions_multiple_values",
+			regions: [][]region{
+				{
+					{
+						b: 0,
+						e: 512000,
+					},
+					{
+						b: 512001,
+						e: 1024000,
+					},
+				},
+				{
+					{
+						b: 0,
+						e: 512000,
+					},
+					{
+						b: 512001,
+						e: 1024000,
+					},
+				},
+				{
+					{
+						b: 0,
+						e: 512000,
+					},
+					{
+						b: 512001,
+						e: 1024000,
+					},
+				},
+			},
+			roundtripCount: 1,
+		},
+		{
+			name: "different_regions",
+			regions: [][]region{
+				{
+					{
+						b: 0,
+						e: 100,
+					},
+				},
+				{
+					{
+						b: 101,
+						e: 200,
+					},
+				},
+				{
+					{
+						b: 201,
+						e: 300,
+					},
+				},
+			},
+			roundtripCount: 3,
+		},
+		{
+			name: "some_overlap",
+			regions: [][]region{
+				{
+					{
+						b: 0,
+						e: 100,
+					},
+				},
+				{
+					{
+						b: 0,
+						e: 100,
+					},
+				},
+				{
+					{
+						b: 201,
+						e: 300,
+					},
+				},
+			},
+			roundtripCount: 2,
+		},
+	}
+
+	var wg sync.WaitGroup
+	// we always run 3 routines
+	routines := 3
+
+	for _, tst := range tests {
+		var (
+			tr = &callsCountRoundTripper{}
+			b  = &blob{
+				fetcher: &fetcher{
+					url: "test",
+					tr:  tr,
+				},
+			}
+		)
+
+		start := make(chan struct{})
+		wg.Add(routines)
+
+		for i := 0; i < routines; i++ {
+			allData := make(map[region]io.Writer)
+			if i < len(tst.regions) {
+				for j := range tst.regions[i] {
+					allData[tst.regions[i][j]] = ioutil.Discard
```

Comment:
Could we add the check for the contents as well?

Reply:
Done.